server/src/main/java/org/opensearch/gateway/remote/ClusterStateChecksum.java

```diff
@@ -0,0 +1,419 @@
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.DiffableStringMap;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.common.io.stream.BufferedChecksumStreamOutput;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+import org.opensearch.core.xcontent.ToXContentFragment;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParseException;
+import org.opensearch.core.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import com.jcraft.jzlib.JZlib;
+
+import static org.opensearch.core.xcontent.XContentParserUtils.ensureExpectedToken;
+
+/**
+ * Stores checksum for all components in cluster state. This will be used to ensure cluster state is same across all nodes in the cluster.
+ */
+public class ClusterStateChecksum implements ToXContentFragment, Writeable {
+
+    private static final String ROUTING_TABLE_CS = "routing_table";
+    private static final String NODES_CS = "discovery_nodes";
+    private static final String BLOCKS_CS = "blocks";
+    private static final String CUSTOMS_CS = "customs";
+    private static final String COORDINATION_MD_CS = "coordination_md";
+    private static final String SETTINGS_MD_CS = "settings_md";
+    private static final String TRANSIENT_SETTINGS_MD_CS = "transient_settings_md";
+    private static final String TEMPLATES_MD_CS = "templated_md";
+    private static final String CUSTOM_MD_CS = "customs_md";
+    private static final String HASHES_MD_CS = "hashes_md";
+    private static final String INDICES_CS = "indices_md";
+    private static final String CLUSTER_STATE_CS = "cluster_state";
+    private static final Logger logger = LogManager.getLogger(ClusterStateChecksum.class);
+
+    long routingTableChecksum;
+    long nodesChecksum;
+    long blocksChecksum;
+    long clusterStateCustomsChecksum;
+    long coordinationMetadataChecksum;
+    long settingMetadataChecksum;
+    long transientSettingsMetadataChecksum;
+    long templatesMetadataChecksum;
+    long customMetadataMapChecksum;
+    long hashesOfConsistentSettingsChecksum;
+    long indicesChecksum;
+    long clusterStateChecksum;
+
+    public ClusterStateChecksum(ClusterState clusterState) {
+        try (
+            BytesStreamOutput out = new BytesStreamOutput();
+            BufferedChecksumStreamOutput checksumOut = new BufferedChecksumStreamOutput(out)
+        ) {
+            clusterState.routingTable().writeTo(checksumOut);
```

Comment:
Should we maintain two BufferedCheckSumStreamOutput?
1. To calculate the individual checkSumCalculation
2. To calculate the overall checkSum so that we don't need to redo it.

Reply:
Maintaining separate streams would occupy more heap memory. I would prefer to keep it this way.

Reply:
My concern is wouldn't we be writing twice one with checksum and the other without, which I believe would be costly. Would prefer we have a checksummed writeTo implementation that writes the checksum alongside?

Reply:
This validation will mostly be used for testing and disabled in production clusters unless we see an issue around cluster state. Overriding writeTo implementation for creating checksum everytime would require changes in all cluster state entities which might be a bit excessive (and maybe wasteful?), given that this is a testing feature only.